enhancements/ingress/add-headers-to-haproxyconfig-via-ingress-controller-or-route-api.md

```diff
@@ -0,0 +1,413 @@
+---
+title: add-headers-to-haproxyconfig-via-ingress-controller-or-route-api
+authors:
+  - "@miheer"
+reviewers:
+  - "@frobware"
+  - "@gcs278"
+  - "@Miciah"
+approvers:
+  - "@frobware"
+  - "@gcs278"
+  - "@Miciah"
+api-approvers:
+  - "@frobware"
+  - "@gcs278"
+  - "@Miciah"
+creation-date: 2020-06-10
+last-updated: 2020-06-10
+tracking-link:
+  - "https://issues.redhat.com/browse/NE-982"
+status: TODO
+see-also:
+- ""
+replaces:
+superseded-by:
+---
+
+# Set HTTP Headers to haproxy.config either via Ingress Contoller or Route API
+
+## Summary
+
+This enhancement extends the IngressController API and Route API to allow the user to set 
+capture of HTTP headers in the haproxy.config file. Using the new API, users may specify a list of 
+HTTP headers either via IngressController API or Route API that should be set or appended.
+
+## Motivation
+
+Certain headers may need to be set for compliance reasons or other reasons. In OpenShift 3, cluster administrators could
+use a custom HAProxy configuration template to set headers, but using a custom template is not permitted in OpenShift 4.
+
+### User Stories
+
+1. As a cluster administrator, I need to configure an IngressController to set/append certain response HTTP headers which are:
+```conf
+Strict-Transport-Security "max-age=31536000; includeSubdomains"
+adding X-XSS-Protection "1; mode=block"
+adding X-Content-Type-Options "nosniff"
+adding X-Frame-Options "SAMEORIGIN"
+```
+in order to satisfy compliance requirements at global level i.e frontend fe_no_sni, frontend public, frontend fe_sni 
+sections of the haproxy.config.
+2. As a user/devops/developer, I need to configure Route to set/append certain response HTTP headers for a backend which 
+   are mentioned above.
+3. To satisfy this use-case, the cluster administrator can set the IngressController's Spec with the header name and values 
+   or Router's `spec.httpSetHeaders` field to set/append the required headers.
+
+### Goals
+
+1. Enable the cluster administrator to specify HTTP headers in the IngressController CR Spec that should get set/append at the global
+   level for all backends which is not application specific.
+2. Enable the developer or user to specify HTTP headers in the Route Spec of Route(which is not a passthrough route) for their application 
+   that should get set/append for sending to application's backends only.
+
+### Non-Goals
+
+1. Admins or users are responsible for providing valid and correct header names values as inputs.
+2. Admission list for headers.
```

Comment:
this is not applicable.

Reply:
> > @Miciah @miheer @alebedev87 I've been going back and forth on a few things on the API PR but I wanted to bring them up here, I have a few concerns about the enhancement which I don't feel have been adequately addressed yet
> > 
> > * The name `Set` could also be described as `Replace`, I saw a comment that this has been discussed but you planned to stick with `Set`. Could we add some detail on why this decision was made, it is beneficial to document alternatives and the discussion around them so that others can review
> > * The lack of `Append` being included in this API, to me, makes it incomplete. I appreciate this wasn't a direct request from the customer, but, if you take a step back and look at the feature as a whole, when people are manipulating headers, there are more than just `Set` and `Delete` actions, typically there is also an `Add` or `Append` action. I don't think the current justification in the enhancement is sufficient for why not to do this. To me it seems like the addition would make this a more complete feature and the support burden/implementation complexity doesn't seem like it would be changed too much. IMO this should either be included for completeness or we should add a good reason why this adds significant complexity
> 
> "add-header" appends an HTTP header field whose name is specified in and whose value is defined by which follows the log-format rules (see Custom Log Format in section 8.2.4). This may be used to send a cookie to a client for example, or to pass some internal information. This rule is not final, so it is possible to add other similar rules. Note that header addition is performed immediately, so one rule might reuse the resulting header from a previous rule.
> 
> "set-header" does the same as "add-header" except that the header name is first removed if it existed. This is useful when passing security information to the server, where the header must not be manipulated by external users.
> 
> So, that is the reason we don't need add-header.
> 
> https://wohugb.gitbooks.io/haproxy/content/config/http-response.html
> 
> https://www.haproxy.com/documentation/hapee/latest/traffic-routing/rewrites/rewrite-responses/

@Miciah i have updated your doubt on this.